src/snapshot_middleware/mod.rs

```diff
@@ -54,89 +58,245 @@ pub fn snapshot_from_vfs(
+        let file_name = path
+            .file_name()
+            .and_then(|n| n.to_str())
+            .with_context(|| format!("file name of {} is invalid", path.display()))?;
 
-        let init_path = path.join("init.server.lua");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
+        match file_name {
+            "init.server.luau" | "init.server.lua" | "init.client.luau" | "init.client.lua"
+            | "init.luau" | "init.lua" | "init.csv" => return Ok(None),
+            _ => {}
         }
 
-        let init_path = path.join("init.client.luau");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
-        }
+        snapshot_from_path(context, vfs, path)
+    }
+}
 
-        let init_path = path.join("init.client.lua");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_lua_init(context, vfs, &init_path);
-        }
+/// Gets an `init` path for the given directory.
+/// This uses an intrinsic priority list and for compatibility,
+/// it should not be changed.
+fn get_init_path<P: AsRef<Path>>(vfs: &Vfs, dir: P) -> anyhow::Result<Option<PathBuf>> {
+    let path = dir.as_ref();
 
-        let init_path = path.join("init.csv");
-        if vfs.metadata(&init_path).with_not_found()?.is_some() {
-            return snapshot_csv_init(context, vfs, &init_path);
-        }
+    let project_path = path.join("default.project.json");
+    if vfs.metadata(&project_path).with_not_found()?.is_some() {
+        return Ok(Some(project_path));
+    }
+
+    let init_path = path.join("init.luau");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
+
+    let init_path = path.join("init.lua");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
 
-        snapshot_dir(context, vfs, path)
+    let init_path = path.join("init.server.luau");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
+
+    let init_path = path.join("init.server.lua");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
+
+    let init_path = path.join("init.client.luau");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
+
+    let init_path = path.join("init.client.lua");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
+
+    let init_path = path.join("init.csv");
+    if vfs.metadata(&init_path).with_not_found()?.is_some() {
+        return Ok(Some(init_path));
+    }
+
+    Ok(None)
+}
+
+/// Gets a snapshot for a path given an InstanceContext and Vfs, taking
+/// user specified sync rules into account.
+fn snapshot_from_path(
```

Comment:
...can we remove this long if-else chain (as well as the one in `Middleware::from_path`), and handle the default syncing behavior as SyncRules instead (perhaps by adding them after we add the user-defined SyncRules in `middleware/project.rs`)? This could be a great opportunity to clean up this module!

Reply:
I considered this, but I was leery of it since it felt rather invasive to propose without any conversation first. If you're okay with it though, I'm more than willing to swap over because it makes things a lot more consistent and easy to change.

Any thoughts on how we should handle `init` files if we do that though?

Reply:
Probably the same way is acceptable? But looking at the code more we can't just add the default rules in `snapshot_project` though... up to you if you want to change it, I just think it'd be nice to have

Reply:
The same way we do now would work, though it'd still involve an if-else chain. After giving it some though, I think I'm open to doing this in the future but right now it'd involve some weirdness, so I'm not sure I want to bloat this PR with it. It's already pretty big.

Reply:
Upon further thought, it turns out this is minimally invasive and actually just makes things cleaner. I've gone ahead and swapped us over to using sync rules for our default behavior, with the exception of `init` files. I imagine we could expose middleware for that too, but it would add a lot more 'default' syncing rules to iterate through so I'm inclined to not do it for the time being.